bokehjs/test/unit/models/widgets/spinner.ts

```diff
@@ -0,0 +1,119 @@
+
+    expect(input.value).to.be.equal("1.36")
+
+    await mousedown_mouseup(button)
+    expect(input.value).to.be.equal("2.36")
+  })
+
+  it("value should be correct after decrementing with step", async () => {
+    const obj = new Spinner({value: 1.36, step: 1, mode: "float", format: "0.00"})
+    const {view} = await display(obj, [500, 400])
+    const button = view.shadow_el.querySelector(".bk-spin-btn-down")!
+    const input = view.shadow_el.querySelector(".bk-input") as HTMLInputElement
+
+    expect(input.value).to.be.equal("1.36")
+
+    await mousedown_mouseup(button)
+    expect(input.value).to.be.equal("0.36")
+  })
+
+  it("value should be correct after incrementing and decrementing with step", async () => {
+    const obj = new Spinner({value: 1.36, step: 1, mode: "float", format: "0.00"})
+    const {view} = await display(obj, [500, 400])
+
+    const increment_button = view.shadow_el.querySelector(".bk-spin-btn-up")!
+    const decrement_button = view.shadow_el.querySelector(".bk-spin-btn-down")!
+    const input = view.shadow_el.querySelector(".bk-input") as HTMLInputElement
+
+    expect(input.value).to.be.equal("1.36")
+
+    await mousedown_mouseup(increment_button)
+    expect(input.value).to.be.equal("2.36")
+
+    await mousedown_mouseup(decrement_button)
+    expect(input.value).to.be.equal("1.36")
+  })
+
+  it("values should be correct when step precision is higher than value's precision", async () => {
+    const obj = new Spinner({value: 2.36, step: 1.364, mode: "float", format: "0.000"})
+    const {view} = await display(obj, [500, 400])
+
+    const increment_button = view.shadow_el.querySelector(".bk-spin-btn-up")!
+    const decrement_button = view.shadow_el.querySelector(".bk-spin-btn-down")!
+    const input = view.shadow_el.querySelector(".bk-input") as HTMLInputElement
+
+    expect(input.value).to.be.equal("2.360")
+
+    await mousedown_mouseup(increment_button)
+    expect(input.value).to.be.equal("3.724")
+
+    await mousedown_mouseup(increment_button)
+    expect(input.value).to.be.equal("5.088")
+
+    await mousedown_mouseup(decrement_button)
+    expect(input.value).to.be.equal("3.724")
+  })
+
+  it("values should be bound by high and low values", async () => {
+    const obj = new Spinner({value: 2.36, step: 1, high: 5, low: 0.5, mode: "float", format: "0.00"})
+    const {view} = await display(obj, [500, 400])
+
+    const increment_button = view.shadow_el.querySelector(".bk-spin-btn-up")!
+    const decrement_button = view.shadow_el.querySelector(".bk-spin-btn-down")!
+    const input = view.shadow_el.querySelector(".bk-input") as HTMLInputElement
+
+    expect(input.value).to.be.equal("2.36")
+
+    await mousedown_mouseup(increment_button)
+    expect(input.value).to.be.equal("3.36")
+
+    await mousedown_mouseup(increment_button)
+    expect(input.value).to.be.equal("4.36")
+
+    await mousedown_mouseup(increment_button)
+    expect(input.value).to.be.equal("5.00")
```

Comment:
Is this the desired behaviour? I think I might have expected that in this case the increment button simply had no effect at all. Perhaps it should be configurable whether a "partial step" is allowed at the bounds (to exactly attain the bounds) or whether the step is strictly respected at all times (even if it means the bounds values can never be attained)

Same questions for lower bound, of course.

cc @bokeh/dev

Reply:
in-case of the presence of a reach_bounds property and set to true, if a partial step is made to reach the bound exactly, should this partial step be preserved in the followed opposite step? or should the full step be applied?
ex:
high: 5 step:0.5 value 4.7
increment -> 5
decrement -> 4.7 or 4.5?

Reply:
@bokeh/dev thoughts on a preference? It could be made configurable I suppose, but I'd rather just pick a good default to start and only add a configuration if it becomes needed.